boefjes/boefjes/plugins/kat_rpki/main.py

```diff
@@ -13,78 +14,155 @@
-        with RPKI_META_PATH.open() as json_meta_file:
-            rpki_meta = json.load(json_meta_file)
+        rpki_json = load_json(RPKI_PATH)
+        rpki_meta = load_json(RPKI_META_PATH)
+
+    if not BGP_PATH.exists() or cache_out_of_date(BGP_META_PATH):
+        bgp_data, bgp_meta = refresh_cache(
+            BGP_SOURCE_URL, BGP_PATH, BGP_META_PATH, hash_algorithm, file_extension="jsonl"
+        )
+    else:
+        bgp_data = load_jsonl(BGP_PATH)
+        bgp_meta = load_json(BGP_META_PATH)
+
     exists = False
-    notexpired = False
     roas = []
-    for roa in rpki_json["roas"]:
-        if IPAddress(ip) in IPNetwork(roa["prefix"]):
-            exists = True
-            expires = datetime.fromtimestamp(roa["expires"])
-            roas.append({"prefix": roa["prefix"], "expires": expires.strftime("%Y-%m-%dT%H:%M"), "ta": roa["ta"]})
-            if expires > now:
-                notexpired = True
+    bgp_entries = []
+    invalid_bgp_entries = []
+    if isinstance(rpki_json, dict) and "roas" in rpki_json:
+        for roa in rpki_json["roas"]:
+            if IPAddress(ip) in IPNetwork(roa["prefix"]):
+                expires = datetime.fromtimestamp(roa["expires"])
+                asn = roa["asn"]
+                roas.append(
+                    {
+                        "prefix": roa["prefix"],
+                        "expires": expires.strftime("%Y-%m-%dT%H:%M%z"),
+                        "ta": roa["ta"],
+                        "asn": asn,
+                    }
+                )
+                exists = True
+
+                # check validity through bgp json
+                invalid_exists = False
+                for entry in bgp_data:
+                    if IPAddress(ip) in IPNetwork(entry["CIDR"]):
+                        bgp_entries.append(entry)
+                        if entry["ASN"] != asn:
+                            invalid_exists = True
 
-    results = {"vrps_records": roas, "notexpired": notexpired, "exists": exists}
+                if invalid_exists:
+                    invalid_bgp_entries.append({"ip": ip, "asn": asn})
+
+    results = {
+        "vrps_records": roas,
+        "exists": exists,
+        "bgp_entries": bgp_entries,
+        "invalid_bgp_entries": invalid_bgp_entries,
+    }
 
     return [
-        (set(), json.dumps(results)),
+        ({"rpki/results"}, json.dumps(results)),
         (
             {"rpki/cache-meta"},
             json.dumps(rpki_meta),
         ),
+        (
+            {"rpki/bgp-cache-meta"},
+            json.dumps(bgp_meta),
+        ),
     ]
 
 
-def create_hash(data: bytes, algo: str) -> str:
-    hashfunc = getattr(hashlib, algo)
-    return hashfunc(data).hexdigest()
-
-
-def cache_out_of_date() -> bool:
-    """Returns True if the file is older than the allowed cache_timout"""
+def cache_out_of_date(meta_path: Path) -> bool:
+    """Returns True if the cache file is older than the allowed cache_timeout"""
     now = datetime.utcnow()
```

Comment:
https://sonarcloud.io/project/issues?resolved=false&id=minvws_nl-kat-coordination&open=AY_dVgxOGgO8mB8X8BEZ
datetime.utcnow does not yield a timezone aware datetime object, changing this means we'd also need to compare against a timezone aware datetime later on, but it is cleaner to do so.

Reply:
I think the easiest way to do that is to use `isoformat` and `fromisoformat` methods that will also write the timezone information:

```
In [2]: now = datetime.now(timezone.utc).isoformat()

In [3]: now
Out[3]: '2024-07-04T09:14:56.348539+00:00'

In [4]: datetime.fromisoformat(now)
Out[4]: datetime.datetime(2024, 7, 4, 9, 14, 56, 348539, tzinfo=datetime.timezone.utc)
```

Reply:
Since we just use `now` to check against max-age `return (now - cached_file_timestamp).total_seconds() > maxage`, would this make any difference?

Reply:
Technically, no. As both are as timezone unaware as each other, but for cleanness it would be better to specify which timezone we are dealing with, especially if we end up comparing a timestamp from an external source against any of these at some point.

Reply:
I actually agree with you haha. I have added timezone awareness with backward compatibility.